src/core_plugins/elasticsearch/lib/cluster.js

```diff
@@ -0,0 +1,133 @@
+import elasticsearch from 'elasticsearch';
+import { get, set, isEmpty, cloneDeep, pick } from 'lodash';
+import toPath from 'lodash/internal/toPath';
+import Boom from 'boom';
+
+import filterHeaders from './filter_headers';
+import { parseConfig } from './parse_config';
+
+export default class Cluster {
+  constructor(config) {
+    this._config = Object.assign({}, config);
+    this.errors = elasticsearch.errors;
+
+    this._client = this.createClient();
+    this._noAuthClient = this.createClient({ auth: false });
+
+    return this;
+  }
+
+  /**
+   * callWithRequest
+   *
+   * Performs a call to ES, passing through whitelisted headers in the request
+   *
+   * @param {Object|undefined} req - The request object
+   * @param {string} endpoint
+   * @param {Object} clientParams
+   * @param {Object} options
+   * @param {boolean} options.wrap401Errors
+   * @returns {Promise}
+   */
+  callWithRequest = (req = {}, endpoint, clientParams = {}, options = {}) => {
+    if (req.headers) {
+      const filteredHeaders = filterHeaders(req.headers, this.getRequestHeadersWhitelist());
+      set(clientParams, 'headers', filteredHeaders);
+    }
+
+    return callAPI(this._noAuthClient, endpoint, clientParams, options);
+  }
+
+  /**
+   * callWithInternalUser
+   *
+   * Performs a call to ES using the credentials in kibana.yml
+   *
+   * @param {string} endpoint
+   * @param {Object} clientParams
+   * @param {Object} options
+   * @returns {Promise}
+   */
+
+  callWithInternalUser = (endpoint, clientParams = {}, options = {}) => {
+    return callAPI(this._client, endpoint, clientParams, options);
+  }
+
+  getRequestHeadersWhitelist = () => getClonedProperty(this._config, 'requestHeadersWhitelist');
+
+  getCustomHeaders = () => getClonedProperty(this._config, 'customHeaders');
+
+  getRequestTimeout = () => getClonedProperty(this._config, 'requestTimeout');
+
+  getUrl = () => getClonedProperty(this._config, 'url');
+
+  getSsl = () => getClonedProperty(this._config, 'ssl');
+
+  getClient = () => this._client;
```

Comment:
This doesn't matter, and I'm not advocating for changing it, but I am curious: why not use getters here?